File: 3661. maximum-walls-destroyed-by-robots.java
class Solution {
    public int maxWalls(int[] robots, int[] distance, int[] walls) {
        int n = robots.length;
        int[][] arr = new int[n][2];
        
        for (int i = 0; i < n; i++) {
            arr[i][0] = robots[i];
            arr[i][1] = distance[i];
        }
        
        Arrays.sort(arr, (a, b) -> Integer.compare(a[0], b[0]));
        Arrays.sort(walls);
        
        int exactWalls = 0;
        for (int i = 0; i < n; i++) {
            if (Arrays.binarySearch(walls, arr[i][0]) >= 0) {
                exactWalls++;
            }
        }
        
        int[][] dp = new int[n][2];
        dp[0][0] = countW(walls, arr[0][0] - arr[0][1], arr[0][0] - 1);
        dp[0][1] = 0;
        
        for (int i = 1; i < n; i++) {
            int pPrev = arr[i - 1][0];
            int dPrev = arr[i - 1][1];
            int pCurr = arr[i][0];
            int dCurr = arr[i][1];
            
            int aStart = pPrev + 1;
            int aEnd = Math.min(pCurr - 1, pPrev + dPrev);
            int cA = countW(walls, aStart, aEnd);
            
            int bStart = Math.max(pPrev + 1, pCurr - dCurr);
            int bEnd = pCurr - 1;
            int cB = countW(walls, bStart, bEnd);
            
            int cBoth = countW(walls, bStart, aEnd);
            int unionCount = cA + cB - cBoth;
            
            dp[i][0] = Math.max(dp[i - 1][0] + cB, dp[i - 1][1] + unionCount);
            dp[i][1] = Math.max(dp[i - 1][0], dp[i - 1][1] + cA);
        }
        
        int lastRight = countW(walls, arr[n - 1][0] + 1, arr[n - 1][0] + arr[n - 1][1]);
        int maxDestroyed = Math.max(dp[n - 1][0], dp[n - 1][1] + lastRight);
        
        return maxDestroyed + exactWalls;
    }
    
    private int countW(int[] arr, int l, int r) {
        if (l > r) return 0;
        int leftIdx = getL(arr, l);
        int rightIdx = getU(arr, r) - 1;
        return Math.max(0, rightIdx - leftIdx + 1);
    }
    
    private int getL(int[] arr, int target) {
        int l = 0, r = arr.length - 1;
        int ans = arr.length;
        while (l <= r) {
            int mid = l + (r - l) / 2;
            if (arr[mid] >= target) {
                ans = mid;
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        return ans;
    }
    
    private int getU(int[] arr, int target) {
        int l = 0, r = arr.length - 1;
        int ans = arr.length;
        while (l <= r) {
            int mid = l + (r - l) / 2;
            if (arr[mid] > target) {
                ans = mid;
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        return ans;
    }
    
    // Test cases
    public static void main(String[] args) {
        Solution solution = new Solution();
        
        // Example 1
        int[] robots1 = {4};
        int[] distance1 = {3};
        int[] walls1 = {1, 10};
        int result1 = solution.maxWalls(robots1, distance1, walls1);
        System.out.println("Example 1: Input: robots = [4], distance = [3], walls = [1, 10]");
        System.out.println("Output: " + result1);
        System.out.println("Expected: 1\n");
        
        // Example 2
        int[] robots2 = {10, 2};
        int[] distance2 = {5, 1};
        int[] walls2 = {5, 2, 7};
        int result2 = solution.maxWalls(robots2, distance2, walls2);
        System.out.println("Example 2: Input: robots = [10, 2], distance = [5, 1], walls = [5, 2, 7]");
        System.out.println("Output: " + result2);
        System.out.println("Expected: 3\n");
        
        // Example 3
        int[] robots3 = {1, 2};
        int[] distance3 = {100, 1};
        int[] walls3 = {10};
        int result3 = solution.maxWalls(robots3, distance3, walls3);
        System.out.println("Example 3: Input: robots = [1, 2], distance = [100, 1], walls = [10]");
        System.out.println("Output: " + result3);
        System.out.println("Expected: 0");
    }
}

File: 3661. maximum-walls-destroyed-by-robots.py
class Solution:
    def maxWalls(self, robots: List[int], distance: List[int], walls: List[int]) -> int:
        # Preprocess: deduplicate walls and sort robots by position
        walls = sorted(set(walls))
        bots = sorted(zip(robots, distance))
        n = len(bots)
        
        if n == 0 or not walls:
            return 0
        
        # Calculate attack intervals for each robot
        left_intervals = []
        right_intervals = []
        
        for i, (pos, dist) in enumerate(bots):
            left_boundary = bots[i-1][0] + 1 if i > 0 else float('-inf')
            right_boundary = bots[i+1][0] - 1 if i < n-1 else float('inf')
            
            # Left attack interval: [max(pos-dist, left_boundary), pos]
            left_intervals.append((max(pos - dist, left_boundary), pos))
            
            # Right attack interval: [pos, min(pos+dist, right_boundary)]
            right_intervals.append((pos, min(pos + dist, right_boundary)))
        
        def count_walls(start: int, end: int) -> int:
            if start > end:
                return 0
            return bisect_right(walls, end) - bisect_left(walls, start)
        
        def get_overlap(interval1: tuple, interval2: tuple) -> int:
            start = max(interval1[0], interval2[0])
            end = min(interval1[1], interval2[1])
            return count_walls(start, end) if start <= end else 0
        
        # Calculate walls count for each interval
        left_counts = [count_walls(start, end) for start, end in left_intervals]
        right_counts = [count_walls(start, end) for start, end in right_intervals]
        
        # Dynamic programming: choose left or right for each robot
        dp_left = left_counts[0]
        dp_right = right_counts[0]
        
        for i in range(1, n):
            # Current robot chooses left
            new_left = max(
                dp_left + left_counts[i],  # Previous also chose left
                dp_right + left_counts[i] - get_overlap(left_intervals[i], right_intervals[i-1])
            )
            
            # Current robot chooses right
            new_right = max(
                dp_right + right_counts[i],  # Previous also chose right
                dp_left + right_counts[i] - get_overlap(right_intervals[i], left_intervals[i-1])
            )
            
            dp_left, dp_right = new_left, new_right
        
        return max(dp_left, dp_right)


# Test cases
if __name__ == "__main__":
    from typing import List
    
    solution = Solution()
    
    # Example 1
    robots1 = [4]
    distance1 = [3]
    walls1 = [1, 10]
    result1 = solution.maxWalls(robots1, distance1, walls1)
    print(f"Example 1: Input: robots = {robots1}, distance = {distance1}, walls = {walls1}")
    print(f"Output: {result1}")
    print(f"Expected: 1\n")
    
    # Example 2
    robots2 = [10, 2]
    distance2 = [5, 1]
    walls2 = [5, 2, 7]
    result2 = solution.maxWalls(robots2, distance2, walls2)
    print(f"Example 2: Input: robots = {robots2}, distance = {distance2}, walls = {walls2}")
    print(f"Output: {result2}")
    print(f"Expected: 3\n")
    
    # Example 3
    robots3 = [1, 2]
    distance3 = [100, 1]
    walls3 = [10]
    result3 = solution.maxWalls(robots3, distance3, walls3)
    print(f"Example 3: Input: robots = {robots3}, distance = {distance3}, walls = {walls3}")
    print(f"Output: {result3}")
    print(f"Expected: 0")